contracts/core/liquidators/twap-impl/TWAPAuction.sol

```diff
@@ -29,14 +40,389 @@ import { TwoAssetPriceBoundedLinearAuction } from "../impl/TwoAssetPriceBoundedL
+            _twapAuction.chunkAuction.auction,
+            ISetToken(address(0)),
+            ISetToken(address(0))
+        );
+    }
+
+    /**
+     * Validates that chunk size is within asset bounds and passed chunkAuctionLength
+     * is unlikely to push TWAPAuction beyond failAuctionPeriod.
+     *
+     * @param _currentSet                   The Set to rebalance from
+     * @param _nextSet                      The Set to rebalance to
+     * @param _startingCurrentSetQuantity   Quantity of currentSet to rebalance
+     * @param _liquidatorData               Struct of parsed liquidator data
+     */
+    function validateLiquidatorData(
+        ISetToken _currentSet,
+        ISetToken _nextSet,
+        uint256 _startingCurrentSetQuantity,
+        TWAPLiquidatorData memory _liquidatorData
+    )
+        internal
+        view
+    {
+        uint256 rebalanceVolume = LiquidatorUtils.calculateRebalanceVolume(
+            _currentSet,
+            _nextSet,
+            oracleWhiteList,
+            _startingCurrentSetQuantity
+        );
+
+        bytes32 assetPairHash = getAssetPairHashFromCollateral(_currentSet, _nextSet);
+
+        validateTWAPParameters(
+            _liquidatorData,
+            assetPairHash,
+            rebalanceVolume
+        );
+    }
+
+    /**
+     * Validates passed in parameters for TWAP auction
+     *
+     * @param _twapLiquidatorData       Struct of passed liquidator data
+     * @param _assetPairHash            Hash of asset pair being rebalanceds
+     * @param _rebalanceVolume          Value of collateral being rebalanced
+     */
+    function validateTWAPParameters(
+        TWAPLiquidatorData memory _twapLiquidatorData,
+        bytes32 _assetPairHash,
+        uint256 _rebalanceVolume
+    )
+        internal
+        view
+    {
+        require(
+            chunkSizeWhiteList[_assetPairHash].isWithin(_twapLiquidatorData.usdChunkSize),
+            "TWAPAuction.validateTWAPParameters: Passed chunk size must be between bounds."
+        );
+
+        // Want to make sure that the expected length of the auction is less than the failAuctionPeriod
+        // or else a legitimate auction could be failed. Calculated as such:
+        // expectedTWAPTime = numChunkAuctions * expectedChunkAuctionLength + (numChunkAuctions - 1) *
+        // chunkAuctionPeriod
+        uint256 numChunkAuctions = _rebalanceVolume.divCeil(_twapLiquidatorData.usdChunkSize);
+        uint256 expectedTWAPAuctionTime = numChunkAuctions.mul(expectedChunkAuctionLength)
+            .add(numChunkAuctions.sub(1).mul(_twapLiquidatorData.chunkAuctionPeriod));
+
+        uint256 failAuctionPeriod = IRebalancingSetTokenV3(msg.sender).failAuctionPeriod();
+
+        require(
+            expectedTWAPAuctionTime < failAuctionPeriod,
+            "TWAPAuction.validateTWAPParameters: Expected auction duration exceeds allowed length."
+        );
+    }
+
+    /**
+     * Validates next chunk auction can begin.
```

Comment:
Add a bit more description: `The next chunk auction can begin when the previous auction has completed, there are still currentSets to rebalance, and the auction period has elapsed`